### ISSUE TITLE
PKCS#11 support

### ISSUE THREAD

**Issue by martin-wanderer** [openconnect/openconnect-gui#107]
Hello. Are Smart Cards supported by openconnect in Windows now?
I use command like this
`openconnect --certificate="pkcs11:model=eToken;id=1324" myvpnserver.com`
and in Ubuntu program gracefully asks me for my PIN.
However, in Windows same command gives me an error
> Error loading certificate from PKCS#11: The requested data were not available

In Ubuntu i used `pkcs11.conf` file to identify driver module. Is there way to do the same in Windows?
Thanks in advance.

**Comment by nmav** [comment]
In windows openconnect-gui supports the native API for smart cards (i.e., CNG). If your smart card driver provides that you'll see your key to select in the GUI. For the command line app, this is the wrong place to ask (but it should work as long as gnutls is compiled with p11-kit support and you can configure p11-kit in windows).

**Comment by martin-wanderer** [comment]
Well, I tried command line app only because I didn't find my key in the GUI. Where should it be visible - in System Store tab? BTW, i have eToken PRO 72k Java and manual on relevant eToken PKI Client claims that it supports CNG.

**Comment by nmav** [comment]
Yes the system store tab is the right one. Could it be missing drivers? You can test whether your version of openconnect works with CNG if you have a yubikey.

**Comment by martin-wanderer** [comment]
Ahhh, there's a remark in the manual, that CNG driver has limited functionality. I believe the solution is in configuring p11-kit. I havn't found any reference on using p11-kit in windows neither at p11 project site nor in google. I understand, this isn't the right place for such question, but since you said it can be done, could you possibly give any clue about where to read about it?

**Comment by nmav** [comment]
The best would be to try asking at the p11-kit mailing list. I've never used p11-kit on windows.

**Comment by nmav** [comment]
The UI could still support PKCS#11 directly via gnutls and list the available keys, but I do not know what impact would have that on windows (I'd have expected all smart cards to support CNG), but it would be a nice feature for MacOSX. I'm tagging it as a feature and leaving it open.

Most likely the only thing needed is for the UI to provide a way to specify a PKCS#11 provider (shared object), and all to select from the available keys (similarly to system interface).

https://www.gnutls.org/manual/html_node/Smart-cards-and-HSMs.html
